File: android/src/api/communication.ts
import { get, post } from './client'

export type SenderType =
  | 'BUYER'
  | 'SELLER'
  | 'EMPLOYEE'
  | 'COMMERCE_ADMIN'
  | 'SUPER_ADMIN'
  | 'SYSTEM'

export type NotificationType =
  | 'NEW_ORDER'
  | 'ORDER_ACCEPTED'
  | 'ORDER_REJECTED'
  | 'ORDER_PREPARING'
  | 'ORDER_READY_FOR_PICKUP'
  | 'COURIER_ASSIGNED'
  | 'COURIER_PICKED_UP'
  | 'DELIVERY_IN_TRANSIT'
  | 'COURIER_NEAR_DESTINATION'
  | 'COURIER_ARRIVED'
  | 'DELIVERED'
  | 'BUYER_RECEIPT_REQUIRED'
  | 'ORDER_COMPLETED'
  | 'ORDER_CANCELLED'
  | 'DELIVERY_ASSIGNED'
  | 'DELIVERY_FAILED'
  | 'DELIVERY_DELAYED'
  | 'NEW_MESSAGE'
  | 'PAYMENT_CONFIRMED'
  | 'CASH_CONFIRMATION_REQUIRED'
  | 'NEW_REVIEW'

export interface OrderConversation {
  id: string
  order_id: string
  buyer_id: string
  shop_id: string
  business_id: string
  created_at: string
  updated_at: string
}

export interface OrderMessage {
  id: string
  conversation_id: string
  sender_user_id: string
  sender_type: SenderType
  sender_name: string
  body: string
  is_admin_intervention: boolean
  read_by_buyer_at?: string
  read_by_seller_at?: string
  created_at: string
}

export interface OrderConversationDetail {
  conversation: OrderConversation
  order_number: string
  order_status: string
  delivery_method: string
  final_total: number
  shop_name: string
  business_name: string
  buyer_name: string
  messages: OrderMessage[]
}

export interface ConversationListItem {
  conversation_id: string
  order_id: string
  order_number: string
  order_status: string
  buyer_id: string
  buyer_name: string
  shop_id: string
  shop_name: string
  business_id: string
  business_name: string
  last_message: string
  last_sender_type: SenderType
  last_message_at: string
  unread_count: number
  created_at: string
}

export interface NotificationItem {
  id: string
  type: NotificationType
  title: string
  body: string
  reference_type: string
  reference_id: string
  metadata: Record<string, unknown>
  read_at?: string
  created_at: string
  is_read: boolean
}

export interface UnreadCounts {
  unread_messages: number
  unread_notifications: number
}

// ----------------- Buyer & Seller APIs -----------------

export async function fetchOrderConversation(orderId: string): Promise<OrderConversationDetail> {
  return get<OrderConversationDetail>(`/orders/${orderId}/conversation`)
}

export async function sendOrderMessage(orderId: string, body: string): Promise<OrderMessage> {
  return post<OrderMessage>(`/orders/${orderId}/messages`, { body })
}

export async function fetchBuyerConversations(params?: { limit?: number; offset?: number }): Promise<{
  items: ConversationListItem[]
  total: number
  limit: number
  offset: number
}> {
  const qs = new URLSearchParams()
  if (params?.limit) qs.set('limit', String(params.limit))
  if (params?.offset) qs.set('offset', String(params.offset))
  const q = qs.toString()
  return get<{
    items: ConversationListItem[]
    total: number
    limit: number
    offset: number
  }>(q ? `/buyer/conversations?${q}` : '/buyer/conversations')
}

export async function fetchBuyerUnreadCounts(): Promise<UnreadCounts> {
  return get<UnreadCounts>('/buyer/unread-counts')
}

export async function fetchSellerConversations(params?: {
  shop_id?: string
  business_id?: string
  limit?: number
  offset?: number
}): Promise<{
  items: ConversationListItem[]
  total: number
  limit: number
  offset: number
}> {
  const qs = new URLSearchParams()
  if (params?.shop_id) qs.set('shop_id', params.shop_id)
  if (params?.business_id) qs.set('business_id', params.business_id)
  if (params?.limit) qs.set('limit', String(params.limit))
  if (params?.offset) qs.set('offset', String(params.offset))
  const q = qs.toString()
  return get<{
    items: ConversationListItem[]
    total: number
    limit: number
    offset: number
  }>(q ? `/seller/conversations?${q}` : '/seller/conversations')
}

export async function fetchSellerUnreadCounts(params?: {
  shop_id?: string
  business_id?: string
}): Promise<UnreadCounts> {
  const qs = new URLSearchParams()
  if (params?.shop_id) qs.set('shop_id', params.shop_id)
  if (params?.business_id) qs.set('business_id', params.business_id)
  const q = qs.toString()
  return get<UnreadCounts>(q ? `/seller/unread-counts?${q}` : '/seller/unread-counts')
}

// ----------------- Notification APIs -----------------

export async function fetchNotifications(params?: { limit?: number; offset?: number }): Promise<{
  items: NotificationItem[]
  total: number
  limit: number
  offset: number
}> {
  const qs = new URLSearchParams()
  if (params?.limit) qs.set('limit', String(params.limit))
  if (params?.offset) qs.set('offset', String(params.offset))
  const q = qs.toString()
  return get<{
    items: NotificationItem[]
    total: number
    limit: number
    offset: number
  }>(q ? `/notifications?${q}` : '/notifications')
}

export async function markNotificationRead(id: string): Promise<{ status: string }> {
  return post<{ status: string }>(`/notifications/${id}/read`)
}

export async function markAllNotificationsRead(): Promise<{ status: string }> {
  return post<{ status: string }>('/notifications/read-all')
}

export async function fetchUnreadNotificationsCount(): Promise<{ unread_count: number }> {
  return get<{ unread_count: number }>('/notifications/unread-count')
}

// ----------------- Courier Action APIs -----------------

export async function confirmCourierArrival(orderId: string): Promise<{ message: string; delivery_status: string }> {
  return post<{ message: string; delivery_status: string }>(`/orders/${orderId}/courier-arrived`)
}

export async function confirmCourierPickedUp(orderId: string): Promise<{ message: string; delivery_status: string }> {
  return post<{ message: string; delivery_status: string }>(`/orders/${orderId}/courier-picked-up`)
}

export async function confirmCourierNearDestination(orderId: string): Promise<{ message: string; delivery_status: string }> {
  return post<{ message: string; delivery_status: string }>(`/orders/${orderId}/courier-near-destination`)
}

File: backend/internal/models/communication.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type SenderType string

type RecipientScope string

const (
	RecipientScopeBuyer       RecipientScope = "BUYER"
	RecipientScopeSellerOwner RecipientScope = "SELLER_OWNER"
	RecipientScopeEmployee    RecipientScope = "EMPLOYEE"
	RecipientScopeAll         RecipientScope = "ALL_PARTICIPANTS"
)

const (
	SenderTypeBuyer         SenderType = "BUYER"
	SenderTypeSeller        SenderType = "SELLER"
	SenderTypeEmployee      SenderType = "EMPLOYEE"
	SenderTypeCommerceAdmin SenderType = "COMMERCE_ADMIN"
	SenderTypeSuperAdmin    SenderType = "SUPER_ADMIN"
	SenderTypeSystem        SenderType = "SYSTEM"
)

type NotificationType string

const (
	// Order Phases (Buyer & Seller)
	NotificationTypeNewOrder             NotificationType = "NEW_ORDER"
	NotificationTypeOrderAccepted        NotificationType = "ORDER_ACCEPTED"
	NotificationTypeOrderRejected        NotificationType = "ORDER_REJECTED"
	NotificationTypeOrderPreparing       NotificationType = "ORDER_PREPARING"
	NotificationTypeOrderReady           NotificationType = "ORDER_READY_FOR_PICKUP"
	NotificationTypeOrderReadyForPickup  NotificationType = "ORDER_READY_FOR_PICKUP"
	NotificationTypeOrderCancelled       NotificationType = "ORDER_CANCELLED"
	NotificationTypeOrderCompleted       NotificationType = "ORDER_COMPLETED"
	NotificationTypeBuyerReceiptRequired NotificationType = "BUYER_RECEIPT_REQUIRED"
	// NotificationTypeOrderReceived announces physical receipt only. It is deliberately
	// distinct from ORDER_COMPLETED, which additionally requires verified cash payment.
	NotificationTypeOrderReceived NotificationType = "ORDER_RECEIVED"

	// Delivery & Courier Lifecycle
	NotificationTypeCourierAssigned        NotificationType = "COURIER_ASSIGNED"
	NotificationTypeCourierPickedUp        NotificationType = "COURIER_PICKED_UP"
	NotificationTypeDeliveryInTransit      NotificationType = "DELIVERY_IN_TRANSIT"
	NotificationTypeCourierNearDestination NotificationType = "COURIER_NEAR_DESTINATION"
	NotificationTypeCourierArrived         NotificationType = "COURIER_ARRIVED"
	NotificationTypeOrderDelivered         NotificationType = "DELIVERED"
	NotificationTypeDelivered              NotificationType = "DELIVERED"
	NotificationTypeDeliveryAssigned       NotificationType = "DELIVERY_ASSIGNED"
	NotificationTypeDeliveryPending        NotificationType = "DELIVERY_PENDING_ASSIGNMENT"
	NotificationTypeDeliveryFailed         NotificationType = "DELIVERY_FAILED"
	NotificationTypeDeliveryDelayed        NotificationType = "DELIVERY_DELAYED"

	// Communication, Payment & Reviews
	NotificationTypeNewMessage               NotificationType = "NEW_MESSAGE"
	NotificationTypePaymentConfirmed         NotificationType = "PAYMENT_CONFIRMED"
	NotificationTypeCashConfirmationRequired NotificationType = "CASH_CONFIRMATION_REQUIRED"
	NotificationTypeNewReview                NotificationType = "NEW_REVIEW"
)

type OrderConversation struct {
	ID         uuid.UUID `json:"id" db:"id"`
	OrderID    uuid.UUID `json:"order_id" db:"order_id"`
	BuyerID    uuid.UUID `json:"buyer_id" db:"buyer_id"`
	ShopID     uuid.UUID `json:"shop_id" db:"shop_id"`
	BusinessID uuid.UUID `json:"business_id" db:"business_id"`
	CreatedAt  time.Time `json:"created_at" db:"created_at"`
	UpdatedAt  time.Time `json:"updated_at" db:"updated_at"`
}

type OrderMessage struct {
	ID                  uuid.UUID      `json:"id" db:"id"`
	ConversationID      uuid.UUID      `json:"conversation_id" db:"conversation_id"`
	SenderUserID        uuid.UUID      `json:"sender_user_id" db:"sender_user_id"`
	SenderType          SenderType     `json:"sender_type" db:"sender_type"`
	SenderName          string         `json:"sender_name" db:"sender_name"`
	Body                string         `json:"body" db:"body"`
	IsAdminIntervention bool           `json:"is_admin_intervention" db:"is_admin_intervention"`
	RecipientScope      RecipientScope `json:"recipient_scope" db:"recipient_scope"`
	RecipientUserID     *uuid.UUID     `json:"recipient_user_id,omitempty" db:"recipient_user_id"`
	ReadByBuyerAt       *time.Time     `json:"read_by_buyer_at,omitempty" db:"read_by_buyer_at"`
	ReadBySellerAt      *time.Time     `json:"read_by_seller_at,omitempty" db:"read_by_seller_at"`
	CreatedAt           time.Time      `json:"created_at" db:"created_at"`
}

type Notification struct {
	ID            uuid.UUID              `json:"id" db:"id"`
	UserID        uuid.UUID              `json:"user_id" db:"user_id"`
	Type          NotificationType       `json:"type" db:"type"`
	Title         string                 `json:"title" db:"title"`
	Body          string                 `json:"body" db:"body"`
	ReferenceType string                 `json:"reference_type" db:"reference_type"`
	ReferenceID   uuid.UUID              `json:"reference_id" db:"reference_id"`
	Metadata      map[string]interface{} `json:"metadata" db:"metadata"`
	ReadAt        *time.Time             `json:"read_at,omitempty" db:"read_at"`
	CreatedAt     time.Time              `json:"created_at" db:"created_at"`
}

type SendMessageRequest struct {
	Body string `json:"body" binding:"required"`
}

type AdminInterveneRequest struct {
	Body            string         `json:"body" binding:"required"`
	RecipientScope  RecipientScope `json:"recipient_scope" binding:"required"`
	RecipientUserID *uuid.UUID     `json:"recipient_user_id"`
}

type OrderConversationParticipant struct {
	UserID     uuid.UUID  `json:"user_id"`
	Name       string     `json:"name"`
	Type       string     `json:"type"`
	LastReadAt *time.Time `json:"last_read_at,omitempty"`
}

type OrderConversationDetailResponse struct {
	Conversation   OrderConversation              `json:"conversation"`
	OrderNumber    string                         `json:"order_number"`
	OrderStatus    string                         `json:"order_status"`
	DeliveryMethod string                         `json:"delivery_method"`
	FinalTotal     float64                        `json:"final_total"`
	ShopName       string                         `json:"shop_name"`
	BusinessName   string                         `json:"business_name"`
	BuyerName      string                         `json:"buyer_name"`
	Participants   []OrderConversationParticipant `json:"participants"`
	Messages       []OrderMessage                 `json:"messages"`
}

type ConversationListItemResponse struct {
	ConversationID uuid.UUID  `json:"conversation_id"`
	OrderID        uuid.UUID  `json:"order_id"`
	OrderNumber    string     `json:"order_number"`
	OrderStatus    string     `json:"order_status"`
	BuyerID        uuid.UUID  `json:"buyer_id"`
	BuyerName      string     `json:"buyer_name"`
	ShopID         uuid.UUID  `json:"shop_id"`
	ShopName       string     `json:"shop_name"`
	BusinessID     uuid.UUID  `json:"business_id"`
	BusinessName   string     `json:"business_name"`
	LastMessage    string     `json:"last_message"`
	LastSenderType SenderType `json:"last_sender_type"`
	LastMessageAt  time.Time  `json:"last_message_at"`
	UnreadCount    int        `json:"unread_count"`
	CreatedAt      time.Time  `json:"created_at"`
}

type NotificationResponse struct {
	ID            uuid.UUID              `json:"id"`
	Type          string                 `json:"type"`
	Title         string                 `json:"title"`
	Body          string                 `json:"body"`
	ReferenceType string                 `json:"reference_type"`
	ReferenceID   string                 `json:"reference_id"`
	Metadata      map[string]interface{} `json:"metadata"`
	ReadAt        *time.Time             `json:"read_at,omitempty"`
	CreatedAt     time.Time              `json:"created_at"`
	IsRead        bool                   `json:"is_read"`
}

type UnreadCountsResponse struct {
	UnreadMessages      int `json:"unread_messages"`
	UnreadNotifications int `json:"unread_notifications"`
}

func MetadataToJSON(m map[string]interface{}) (string, error) {
	if m == nil {
		return "{}", nil
	}
	b, err := json.Marshal(m)
	if err != nil {
		return "{}", err
	}
	return string(b), nil
}

func JSONToMetadata(data []byte) map[string]interface{} {
	if len(data) == 0 {
		return make(map[string]interface{})
	}
	var m map[string]interface{}
	if err := json.Unmarshal(data, &m); err != nil {
		return make(map[string]interface{})
	}
	return m
}
